feat: add a plugin to integrate with Good Ghosting

OsamaElHariri opened the pull request:
closes #3337

Add a plugin to integrate with [GoodGhosting](https://goodghosting.com/#/), a DeFi savings game dApp on Polygon

All data is retrieved from the smart contracts, similarly to how the data is currently retrieved on the GoodGhosting web page

Once this plugin is enabled, you can see it in action on this [twitter post](https://twitter.com/goodghosting/status/1405204093254221824)

### Video Demo

https://user-images.githubusercontent.com/23722804/124515484-3a961e80-dde8-11eb-82c8-b9e00d199775.mp4

---
### Added features after testing with GoodGhosting (more info in [this comment](https://github.com/DimensionDev/Maskbook/pull/3570#issuecomment-886037411) ):
- Allow players to join the game, make deposits, and withdraw their earnings.
- Allow players to withdraw early from the game before it ends
- Get the GoodGhosting contract address dynamically by fetching this config file: https://github.com/Good-Ghosting/mask-plugin-integration/blob/master/games.json

septs commented:
plz rename this pr name, thankd

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:
```diff
@@ -0,0 +1,110 @@
+                    <div className={classes.circularData}>
+                        <CircularDataDisplay
+                            header={t('plugin_good_ghosting_all_players_status_winning')}
+                            title={`${playerInfo.winning}`}
```
```suggestion
                            title={playerInfo.winning}
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:
```diff
@@ -0,0 +1,110 @@
+                    <div className={classes.circularData}>
+                        <CircularDataDisplay
+                            header={t('plugin_good_ghosting_all_players_status_waiting')}
+                            title={`${playerInfo.waiting}`}
```
```suggestion
                            title={playerInfo.waiting}
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:
```diff
@@ -0,0 +1,110 @@
+                    </div>
+                </Grid>
+            </Grid>
+            <Grid className={classes.infoRow} container justifyContent={'center'}>
```
```suggestion
            <Grid className={classes.infoRow} container justifyContent="center">
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:
```diff
@@ -0,0 +1,110 @@
+                    <div className={classes.circularData}>
+                        <CircularDataDisplay
+                            header={t('plugin_good_ghosting_all_players_status_ghost')}
+                            title={`${playerInfo.ghosts}`}
```
```suggestion
                            title={playerInfo.ghosts}
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/OtherPlayersView.tsx:
```diff
@@ -0,0 +1,110 @@
+                    <div className={classes.circularData}>
+                        <CircularDataDisplay
+                            header={t('plugin_good_ghosting_all_players_status_dropout')}
+                            title={`${playerInfo.dropouts}`}
```
```suggestion
                            title={playerInfo.dropouts}
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx:
```diff
@@ -0,0 +1,116 @@
+
+    let status = t('plugin_good_ghosting_status_unknown')
+
+    switch (getPlayerStatus(props.info.currentPlayer, props.info.currentSegment)) {
+        case PlayerStatus.Winning:
+            status = t('plugin_good_ghosting_status_winning')
+            break
+        case PlayerStatus.Waiting:
+            status = t('plugin_good_ghosting_status_waiting')
+            break
+        case PlayerStatus.Ghost:
+            status = t('plugin_good_ghosting_status_ghost')
+            break
+        case PlayerStatus.Dropout:
+            status = t('plugin_good_ghosting_status_dropout')
+            break
+    }
```
extract as function

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/TimelineView.tsx:
```diff
@@ -0,0 +1,186 @@
+                                        classes.circleIndicator,
+                                        classes.circleIndicatorFilled,
+                                        classes.rightAligned,
+                                    )}></div>
```
```suggestion
                                    )} />
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/hooks/useGameInfo.ts:
```diff
@@ -0,0 +1,78 @@
+
+        return {
+            segmentPayment,
+            firstSegmentStart: Number.parseInt(firstSegmentStart),
+            currentSegment: Number.parseInt(currentSegment),
+            lastSegment: Number.parseInt(lastSegment),
+            segmentLength: Number.parseInt(segmentLength),
+            numberOfPlayers: Number.parseInt(numberOfPlayers),
```
```suggestion
            firstSegmentStart: Number.parseInt(firstSegmentStart, 10),
            currentSegment: Number.parseInt(currentSegment, 10),
            lastSegment: Number.parseInt(lastSegment, 10),
            segmentLength: Number.parseInt(segmentLength, 10),
            numberOfPlayers: Number.parseInt(numberOfPlayers, 10),
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/utils.ts:
```diff
@@ -0,0 +1,19 @@
+    const mostRecentSegmentPaid = Number.parseInt(player.mostRecentSegmentPaid)
+
+    if (player.withdrawn) return PlayerStatus.Dropout
+    else if (mostRecentSegmentPaid < currentSegment - 1) return PlayerStatus.Ghost
+    else if (mostRecentSegmentPaid === currentSegment - 1) return PlayerStatus.Waiting
+    else if (mostRecentSegmentPaid === currentSegment) return PlayerStatus.Winning
```
```suggestion
    if (mostRecentSegmentPaid < currentSegment - 1) return PlayerStatus.Ghost
    if (mostRecentSegmentPaid === currentSegment - 1) return PlayerStatus.Waiting
    if (mostRecentSegmentPaid === currentSegment) return PlayerStatus.Winning
```

OsamaElHariri commented:
Updated the PR based on the suggestions.

Also, I added a timer on top of the tabs similar to what shows on the GoodGhosting website.

septs commented on packages/maskbook/src/plugins/GoodGhosting/README.md:
```diff
@@ -0,0 +1,18 @@
+# Plugin: dHEDGE
```
```suggestion
# Plugin: Good Ghosting
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx:
```diff
@@ -0,0 +1,116 @@
+        )
+    }
+
+    const getStatus = (player: Player, currentSegment: number) => {
+        switch (getPlayerStatus(player, currentSegment)) {
+            case PlayerStatus.Winning:
+                return t('plugin_good_ghosting_status_winning')
+            case PlayerStatus.Waiting:
+                return t('plugin_good_ghosting_status_waiting')
+            case PlayerStatus.Ghost:
+                return t('plugin_good_ghosting_status_ghost')
+            case PlayerStatus.Dropout:
+                return t('plugin_good_ghosting_status_dropout')
+            default:
+                return t('plugin_good_ghosting_status_unknown')
+        }
+    }
```
plz move this function to this file bottom, thanks

Jack-Works commented on packages/maskbook/src/plugins/PluginUI.ts:
```diff
@@ -9,7 +9,9 @@ export const PluginUI: ReadonlySet<PluginConfig> = plugins
 
 sideEffect.then(() => {
     if (Flags.dhedge_enabled) plugins.add(DHedgePluginDefine)
+    if (Flags.good_ghosting_enabled) plugins.add(GoodGhostingPluginDefine)
```
Hello, @OsamaElHariri thanks for your contribution. Please use the new way to define a plugin, you can see many examples in the plugins folder. `dhedge` is the only last plugin not migrated to the new definition standard and we should not add a new definition here.

Jack-Works replied:
FYI, all new APIs are well documented in packages\plugin-infra\src\types.ts and the new plugin structure can be seen in packages\plugins\example.

OsamaElHariri replied:
I went over the example of how the updated plugin infra looks like and updated the PR. This plugin is now defined using the updated way of handling plugins

Jack-Works reviewed (approved):
Need test from @AinuX and blockchain review from @zhouhanseng or @guanbinrui

septs commented on packages/maskbook/src/plugins/GoodGhosting/utils.ts:
```diff
@@ -30,6 +30,10 @@ export function getNextTimelineIndex(timeline: TimelineEvent[]) {
     return timeline.length - 1
 }
 
+export function getIsEndOfTimeline(timelineIndex: number, timeline: TimelineEvent[]) {
```
why function name not is `isEndOfTimeline`

OsamaElHariri replied:
Updated the function name :+1:

AinuX commented:
Just checked “Game”, “Timeline” and “Everyone” tab under Polygon and other unsupported chains, seems OK to me.

But, since there is only one active game pool and have already started, so I can’t to join the game to check the “Personal” tab, and also have no more game pool to test.

One concern, the plugin will be rendered when any url prefixed with “https://goodghosting.com”, even like url “https://goodghosting.com/#/2#faqs”.

guanbinrui commented on packages/maskbook/src/plugins/GoodGhosting/UI/PreviewCard.tsx:
```diff
@@ -0,0 +1,100 @@
+export function PreviewCard(props: PreviewCardProps) {
+    const { value: info, error, loading, retry } = useGameInfo()
+
+    if (loading) {
+        return <Typography color="textPrimary">Loading...</Typography>
+    } else if (error || !info) {
+        return (
+            <Box display="flex" flexDirection="column" alignItems="center">
+                <Typography color="textPrimary">Something went wrong.</Typography>
+                <Button sx={{ marginTop: 1 }} size="small" onClick={retry}>
+                    Retry
+                </Button>
+            </Box>
+        )
```
```suggestion
    if (loading) {
        return <Typography color="textPrimary">Loading...</Typography>
    } 
    
    if (error || !info) {
        return (
            <Box display="flex" flexDirection="column" alignItems="center">
                <Typography color="textPrimary">Something went wrong.</Typography>
                <Button sx={{ marginTop: 1 }} size="small" onClick={retry}>
                    Retry
                </Button>
            </Box>
        )
```

Jack-Works commented:
hi here some merge conflicts

guanbinrui commented on packages/maskbook/src/_locales/en/messages.json:
```diff
@@ -754,5 +754,44 @@
+    "plugin_good_ghosting_deposits": "Deposits made",
+    "plugin_good_ghosting_total_deposited": "Total deposited",
+    "plugin_good_ghosting_address": "Address",
+    "plugin_good_ghosting_not_a_participant": "Looks like you're not a participant in this game",
```
```suggestion
    "plugin_good_ghosting_not_a_participant": "Looks like you're not a participant in this game。",
```

guanbinrui commented on packages/maskbook/src/_locales/en/messages.json:
```diff
@@ -754,5 +754,44 @@
+    "plugin_good_ghosting_address": "Address",
+    "plugin_good_ghosting_not_a_participant": "Looks like you're not a participant in this game",
+    "plugin_good_ghosting_next_event": "Coming up next",
+    "plugin_good_ghosting_game_end": "This game has ended",
```
```suggestion
    "plugin_good_ghosting_game_end": "This game has ended.",
```

OsamaElHariri replied:
Will keep this one as is since it's a title

---

![image](https://user-images.githubusercontent.com/23722804/127024172-a40904dc-5095-47c6-8a7a-91ba2c9de3c4.png)

guanbinrui commented on packages/maskbook/src/_locales/en/messages.json:
```diff
@@ -754,5 +754,44 @@
+    "plugin_good_ghosting_join_game": "Join Game",
+    "plugin_good_ghosting_make_deposit": "Make Deposit",
+    "plugin_good_ghosting_join_help_text": "Join this savings game! To win, you need to deposit money into the savings pool every round, prior to the round's deadline. Deposit {{segmentPayment}} {{token}} every round for {{roundCount}} rounds to win the game. If other players miss depositing in a round and become ghosts, their earnings will go to you.",
+    "plugin_good_ghosting_deposit_help_text": "Deposit {{segmentPayment}} {{token}} for this round and continue to be a winner",
```
```suggestion
    "plugin_good_ghosting_deposit_help_text": "Deposit {{segmentPayment}} {{token}} for this round and continue to be a winner.",
```

OsamaElHariri commented:
After testing with GoodGhosting and discussing the use case of this plugin, it became clear that a couple of more features are needed. The features added are the following (I'll also add these to the PR description above):
- Allow players to join the game, make deposits, and withdraw their earnings.
- Allow players to withdraw early from the game before it ends
- Get the GoodGhosting contract address dynamically by fetching this config file: https://github.com/Good-Ghosting/mask-plugin-integration/blob/master/games.json

Here are some images that show the new additions in the UI:

**Joining a game (make deposit and withdraw actions look similair to this):**
![image](https://user-images.githubusercontent.com/23722804/126866489-269ba6f2-1cc1-4025-b9b1-4d9f2d10a76b.png)
![image](https://user-images.githubusercontent.com/23722804/126866498-74e38a43-daaa-4917-a18d-adcb61a650af.png)


**Withdraw early:**
![image](https://user-images.githubusercontent.com/23722804/126866574-3435c1f8-a734-4ae1-9e19-a3e3d37d3971.png)

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx:
```diff
@@ -0,0 +1,107 @@
+
+interface TimerProps {
+    targetDate: Date
+    updateTargetDate: () => void
```
```suggestion
    onUpdateTargetDate: () => void
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/GameStatsView.tsx:
```diff
@@ -0,0 +1,174 @@
+import { formatBalance } from '@masknet/web3-shared'
+import { makeStyles, Grid, Typography, Box, Button } from '@material-ui/core'
+import { addSeconds, differenceInDays, formatDuration } from 'date-fns/esm'
```
plz split to multi import statement

CC @Jack-Works

OsamaElHariri replied:
Split the date-fns imports to individual statements :+1:

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/GameStatsView.tsx:
```diff
@@ -0,0 +1,174 @@
+        )
+    }
+
+    const getReadableInterval = (roundLength: number) => {
```
move this function to this file bottom

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx:
```diff
@@ -0,0 +1,154 @@
+        setButtonEnabled(false)
+        try {
+            await earlyWithdraw()
+        } catch (error) {
```
plz note why need ignore error

septs replied:
```suggestion
        } catch (error) {
            // ignore error
            // HACK: ....
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/PersonalView.tsx:
```diff
@@ -0,0 +1,154 @@
+    )
+}
+
+function useGetStatus(info: GoodGhostingInfo, player?: Player) {
```
```suggestion
function usePlayerStatusMessage(info: GoodGhostingInfo, player?: Player) {
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx:
```diff
@@ -0,0 +1,107 @@
+import { Grid, makeStyles, Typography } from '@material-ui/core'
+import { intervalToDuration, isBefore } from 'date-fns'
```
plz split to multi import statement

septs replied:
```suggestion
import intervalToDuration from 'date-fns/intervalToDuration'
import isBefore from 'date-fns/isBefore'
```

septs reviewed (changes requested):
plz resolve all warnings

OsamaElHariri commented:
Thanks for the review @septs 

I split the date imports into individual statements, fixed the function names and locations, handled the error instead of ignoring, and fixed the pipeline linter warnings.

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx:
```diff
@@ -0,0 +1,108 @@
+import isBefore from 'date-fns/isBefore'
+import intervalToDuration from 'date-fns/intervalToDuration'
+import { formatDuration } from 'date-fns/esm'
+import { useEffect, useState } from 'react'
```
?

septs commented on packages/maskbook/src/plugins/GoodGhosting/UI/TimelineTimer.tsx:
```diff
@@ -0,0 +1,108 @@
+import { Grid, makeStyles, Typography } from '@material-ui/core'
+import isBefore from 'date-fns/isBefore'
+import intervalToDuration from 'date-fns/intervalToDuration'
+import { formatDuration } from 'date-fns/esm'
```
```suggestion
import formatDuration from 'date-fns/formatDuration'
```

septs commented on packages/maskbook/src/plugins/GoodGhosting/types.tsx:
```diff
@@ -0,0 +1,46 @@
+    refresh: () => void
+}
+
+export interface Player {
+    addr: string
+    amountPaid: string
+    canRejoin: Boolean
+    mostRecentSegmentPaid: string
+    withdrawn: Boolean
+}
```
```suggestion
export interface Player {
    addr: string
    amountPaid: string
    canRejoin: boolean
    mostRecentSegmentPaid: string
    withdrawn: boolean
}
```

Jack-Works commented:
Hi please merge with develop branch, I have moved `createPluginMessage` and `createPluginRPC` into the `@masknet/plugin-infra` package.

OsamaElHariri commented:
I applied the changes and rebased with develop

septs reviewed (approved):
approve, but plz wait for @Jack-Works and @guanbinrui re-check all files

guanbinrui reviewed (approved):
LGTM

septs commented:
@OsamaElHariri plz wait for @guanbinrui merge this branch to develop
